src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -1,12 +1,81 @@
+    ///
+    /// Tokens with target cancelation time 15ms apart would resolve to the same instance.
+    /// </summary>
+    internal static class TimeoutTokenSource
+    {
+        private const int COALESCING_SPAN_MS = 15;
+        private static readonly ConcurrentDictionary<long, Task<CancellationToken>> s_tokenCache =
+            new ConcurrentDictionary<long, Task<CancellationToken>>();
+
+        public static CancellationToken FromTimeout(int millisecondsTimeout)
+        {
+            return FromTimeoutAsync(millisecondsTimeout).Result;
+        }
+
+        public static Task<CancellationToken> FromTimeoutAsync(int millisecondsTimeout)
+        {
+            // Note that CancellationTokenSource constructor requires input to be >= -1,
+            // restricting millisecondsTimeout to be >= -1 would enforce that
+            if (millisecondsTimeout < -1)
+            {
+                throw new ArgumentOutOfRangeException("Invalid millisecondsTimeout value " + millisecondsTimeout);
+            }
+
+            uint currentTime = (uint)Environment.TickCount;
+            long targetTime = millisecondsTimeout + currentTime;
+            // round the targetTime up to the next closest 15ms
+            targetTime = ((targetTime + (COALESCING_SPAN_MS - 1)) / COALESCING_SPAN_MS) * COALESCING_SPAN_MS;
+
+            Task<CancellationToken> tokenTask;
+
+            if (!s_tokenCache.TryGetValue(targetTime, out tokenTask))
+            {
+                var tcs = new TaskCompletionSource<CancellationToken>();
+
+                // only a single thread may succeed adding its task into the cache
+                if (s_tokenCache.TryAdd(targetTime, tcs.Task))
+                {
+                    // Since this thread was successful reserving a spot in the cache, it would be the only thread
+                    // that construct the CancellationTokenSource
+                    var token = new CancellationTokenSource((int)(targetTime - currentTime)).Token;
+
+                    // Clean up cache when Token is canceled
+                    token.Register(t => {
+                        Task<CancellationToken> ignored;
+                        s_tokenCache.TryRemove((long)t, out ignored);
+                    }, targetTime);
+
+                    // set the result so other thread may observe the token, and return
+                    tcs.TrySetResult(token);
+                    tokenTask = tcs.Task;
+                }
+                else
+                {
+                    // for threads that failed when calling TryAdd, there should be one already in the cache
+                    // Unless in unlikely scenario the token was already cancelled and timed out
+                    if (!s_tokenCache.TryGetValue(targetTime, out tokenTask))
+                    {
+                        tokenTask = Task.FromResult(new CancellationTokenSource((int)(targetTime - currentTime)).Token);
```

Comment:
Change to
```csharp
    tokenTask = Task.FromResult(new CancellationTokenSource(millisecondsTimeout).Token);
```
Because we're creating a non-coalesced CT there's no need to round it and we might as well give an accurate Token.